ext/mbstring/mbstring.c

```diff
@@ -2941,6 +2941,130 @@ PHP_FUNCTION(mb_strtolower)
 	RETURN_STR(mbstring_convert_case(PHP_UNICODE_CASE_LOWER, ZSTR_VAL(str), ZSTR_LEN(str), enc));
 }
 
+static bool is_trim_wchar(uint32_t w, uint32_t *characters, size_t characters_len)
+{
+	for (size_t i = 0; i < characters_len; i++) {
+		if (w == characters[i]) {
```

Comment:
Uum.. I think this problem, but I couldn't find a way.

Reply:
Perhaps the distinction between 1 character and >1 character shouldn't be made.

Anyway, I still think a HashTable would be a good idea, combined with avoiding decoding the `what` string repeatedly. This would improve performance.

To do so:
- Create a HashTable variable on the stack
- Use `zend_hash_init` to initialize that HashTable; and `zend_hash_destroy` to destroy it.
- Fill the HashTable with the unicode codepoints of `what` as keys and an empty value as value using `zend_hash_index_add_new`. Do this once before the loop.
- Pass the HashTable instead of a string to `trim_each_wchar`
- In `is_trim_wchar` you can use `zend_hash_index_exists` to check if the unicode codepoint is in the HashTable. This takes constant time so it's much faster than a loop.
- In `trim_each_wchar`  it's not necessary anymore to repeatedly decode `what`, because the HashTable now takes that role.

Note that it's then no longer necessary too to encode the `trim_default_chars` array in `mb_trim_default_chars` to a zend_string. Instead you can fill in the HashTable directly with a for loop.

Let me know if you need help :slightly_smiling_face:

Reply:
Oh, test failed when enable JIT. What's happened?